market_maker_keeper/leverjfutures_market_maker_keeper_v2.py

```diff
@@ -0,0 +1,331 @@
+        self.target_price_lean = Wad(0)
+
+        self.history = History()
+
+        self.leverj_api = LeverjFuturesAPI(web3=self.web3,
+                                    api_server=self.arguments.leverj_api_server,
+                                    account_id=self.arguments.account_id,
+                                    api_key=self.arguments.api_key,
+                                    api_secret=self.arguments.api_secret,
+                                    timeout=self.arguments.leverj_timeout)
+
+
+        self.order_book_manager = OrderBookManager(refresh_frequency=self.arguments.refresh_frequency)
+        self.order_book_manager.get_orders_with(lambda: self.leverj_api.get_orders(self.pair()))
+        self.order_book_manager.get_balances_with(lambda: self.leverj_api.get_balances())
+        self.order_book_manager.cancel_orders_with(lambda order: self.leverj_api.cancel_order(order.order_id))
+        self.order_book_manager.enable_history_reporting(self.order_history_reporter, self.our_buy_orders,
+                                                         self.our_sell_orders)
+        self.order_book_manager.start()
+
+    def main(self):
+        with Lifecycle() as lifecycle:
+            lifecycle.initial_delay(1)
+            lifecycle.on_startup(self.startup)
+            lifecycle.every(1, self.synchronize_orders)
+            lifecycle.on_shutdown(self.shutdown)
+
+    def startup(self):
+        quote_increment = self.leverj_api.get_tickSize(self.pair())
+        self.precision = -(int(log10(float(quote_increment)))+1)
+
+    def shutdown(self):
+        self.order_book_manager.cancel_all_orders()
+
+    def pair(self):
+        name_to_id_map = {'BTCDAI': '1', 'ETHDAI': '2'}
+        return name_to_id_map[self.arguments.pair.upper()]
+
+    def leverage(self):
+        leverage_in_float = self.arguments.leverage
+        if leverage_in_float is None:
+            leverage_in_float = 1.0
+        return leverage_in_float
+
+    def token_sell(self) -> str:
+        return self.arguments.pair.upper()[:3]
+
+    def token_buy(self) -> str:
+        return self.arguments.pair.upper()[3:]
+
+    def allocated_balance(self, token: str) -> Wad:
+        quote_asset_address = self.leverj_api.get_product(self.pair())["quote"]["address"]
+        # for perpetual contracts, the quote balance is allocated across instruments and sides to enter into trades
+        total_available = self.leverj_api.get_plasma_balance(quote_asset_address)
+        # adjust for leverage
+        total_available = Decimal(total_available)*Decimal(self.leverage())
+        self.logger.debug(f'total_available: {total_available}')
+        return self._allocate_to_pair(total_available).get(token)
+
+    def _allocate_to_pair(self, total_available):
+        # total number of instruments across which the total_available balance is distributed
+        # total_available is denominated in quote units
+        total_number_of_instruments = 1
```

Comment:
Would this ever change?

Reply:
We initially tried to trade multiple pairs using same account and then scaled back to 1 account for 1 pair.